### PR TITLE
[SPARK-26520][SQL] data source v2 API refactor (micro-batch read)

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Following https://github.com/apache/spark/pull/23086, this PR does the API refactor for micro-batch read, w.r.t. the [doc](https://docs.google.com/document/d/1uUmKCpWLdh9vHxP7AWJ9EgbwB_U6T3EJYNjhISGmiQg/edit?usp=sharing)

The major changes:
1. rename `XXXMicroBatchReadSupport` to `XXXMicroBatchStream`
2. implement `TableProvider`, `Table`, `ScanBuilder` and `Scan` for streaming sources
3. at the beginning of micro-batch streaming execution, convert `StreamingRelationV2` to `StreamingDataSourceV2Relation` directly, instead of `StreamingExecutionRelation`.

followup:
support operator pushdown for stream sources

## How was this patch tested?

existing tests

### CODE REVIEW

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala [hunk 1]
```diff
@@ -432,6 +400,54 @@ private[kafka010] class KafkaSourceProvider extends DataSourceRegister
...
+    override def name(): String = s"Kafka ${strategy(options.asMap().asScala.toMap)}"
+
+    override def schema(): StructType = KafkaOffsetReader.kafkaSchema
+
+    override def newScanBuilder(options: DataSourceOptions): ScanBuilder = {
+      new KafkaScanBuilder(options)
+    }
+  }
+
+  class KafkaScanBuilder(options: DataSourceOptions) extends ScanBuilder with Scan {
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala]: I don't have a strong preference, currently all the streaming sources don't support operator pushdown, so it's easier to implement both of them together.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/SupportsMicroBatchRead.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * An empty mix-in interface for {@link Table}, to indicate this table supports streaming scan with
+ * micro-batch mode.
+ * <p>
+ * If a {@link Table} implements this interface, the
+ * {@link SupportsRead#newScanBuilder(DataSourceOptions)} must return a {@link ScanBuilder} that
+ * builds {@link Scan} with {@link Scan#toMicroBatchStream()} implemented.
+ * </p>
+ */
+@Evolving
+public interface SupportsMicroBatchRead extends SupportsRead { }
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/SupportsMicroBatchRead.java]: We need to know the capability at the table level. It's too late to do it at the scan level, as creating a scan may be expensive.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/Scan.java [hunk 3]
```diff
@@ -65,4 +67,20 @@ default String description() {
...
+   * implementation, if the {@link Table} that creates this scan implements
+   * {@link SupportsMicroBatchRead}.
+   *
+   * @param checkpointLocation a path to Hadoop FS scratch space that can be used for failure
+   *                           recovery. Data streams for the same logical source in the same query
+   *                           will be given the same checkpointLocation.
+   *
+   * @throws UnsupportedOperationException
+   */
+  default MicroBatchStream toMicroBatchStream(String checkpointLocation) {
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/Scan.java]: Another idea is to merge `Scan` and `Batch/Stream`, see the alternative 1 in the [doc](https://docs.google.com/document/d/1uUmKCpWLdh9vHxP7AWJ9EgbwB_U6T3EJYNjhISGmiQg/edit#heading=h.vwv2g2bnqluq)

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/streaming/SparkDataStream.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import org.apache.spark.sql.execution.streaming.BaseStreamingSource;
+
+/**
+ * The base interface representing a readable data stream in a Spark streaming query. It's
+ * responsible to manage the offsets of the streaming source in the streaming query.
+ *
+ * Data sources should implement concrete data stream interfaces: {@link MicroBatchStream}.
+ */
+@Evolving
+public interface SparkDataStream extends BaseStreamingSource {
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/streaming/SparkDataStream.java]: This is for `ContinuousStream` which will be added later. Please refer to the [doc](https://docs.google.com/document/d/1uUmKCpWLdh9vHxP7AWJ9EgbwB_U6T3EJYNjhISGmiQg/edit#) for more details.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketMicroBatchStream.scala [hunk 5]
```diff
@@ -220,43 +192,3 @@ class TextSocketMicroBatchReadSupport(options: DataSourceOptions)
 }
 
 case class TextSocketInputPartition(slice: ListBuffer[(UTF8String, Long)]) extends InputPartition
-
-class TextSocketSourceProvider extends DataSourceV2
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketMicroBatchStream.scala]: This is moved to a new file: https://github.com/apache/spark/pull/23430/files#diff-c223080834806efd92efc860656b60cdR34

**Comment by cloud-fan** [comment]
cc @jose-torres @tdas @gatorsmile @rdblue @gengliangwang

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/Scan.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (arunmahadevan) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/Scan.java]: In "alternative1" there is no equivalent Logical `Scan`? I was thinking we need the `Scan` (the logical scan) separate from physical scans.

Also if they don't inherit a common parent can it be passed to the DatasourceV2ScanExec ? 

Anyways better to relook and rename as appropriate to keep the different ones (batch/micro-batch/continuous) have common pre/suffixes and denote what they mean.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/streaming/SparkDataStream.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (arunmahadevan) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/streaming/SparkDataStream.java]: Makes sense. A Stream (a stream of events) and Source typically imply different things. In this case the `SparkDataStream` looks to bew more of a source specific thing than a stream.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/SupportsMicroBatchRead.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (arunmahadevan) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/SupportsMicroBatchRead.java]: Only thing is it doesn't enforce anything. A method like `Table.supportedTypes()` might also work.

[the remark this answers is not in the document]
Comment (rdblue) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/SupportsMicroBatchRead.java]: @arunmahadevan, that's similar to the capabilities that we plan to add. Spark will query specific capabilities for a table to make determinations like this to cut down on the number of empty interfaces.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Review by jose-torres — COMMENTED** [review]
LGTM. Note that I haven't really thought about the naming, just tried to confirm that it follows the community consensus in the doc.

**Pull request closed without merging**